### ISSUE TITLE
Can't negate permissions with PlayerHooks.PlayerPermission

### ISSUE THREAD

**Issue by deadsurgeon42** [Pryaxis/TShock#1470]
```cs
    public static void PlayerPermission(PlayerPermissionEventArgs e)
    {
      if (e.Handled) return;

      // result:
      // true = player does have permission, don't let tshock handle it
      // null = idk, let tshock handle it
      // false = player has the permission negated, let tshock know player doesn't have permission (but how)
      bool? result = UserInfo.HasPermission(e.Player, e.Permission);

      if (result == null)
        return;

      e.Handled = true;
      // e.Negated = (bool) result; (maybe?)
    }
```

Right now `PlayerHooks.PlayerPermission` is limited as there's only two ways of telling TShock what to do:

`e.Handled = true` 🡲 The player has permission, don't check or prevent the player
`e.Handled = false` 🡲 I don't know if the player has permission or not, check
`???` 🡲 *The player doesn't have permission, don't check but prevent the player*

This is good enough for adding permissions, but when you want to negate permissions from the player it falls short.

A simple solution could be changing `PlayerHooks.OnPlayerPermission` return type to `bool?` and adding a `bool Negated` property into `PlayerPermissionEventArgs` (and their cousins `Player(Item/Proj/Tile)banPermissionEventArgs`) which is checked if `Handled` is `false`:

```cs
var args = new PlayerPermissionEventArgs(player, permission);
PlayerPermission(args);
if (!args.Handled)
  return !args.Negated;
else
  return null;
```

Which is then checked in `TSPlayer.HasPermission`:

```cs
public bool HasPermission(string permission)
{
  bool? hookResult = PlayerHooks.OnPlayerPermission(this, permission);
  if (hookResult != null)
   return (bool) hookResult;

   if (tempGroup != null)
     return tempGroup.HasPermission(permission);
   else
     return Group.HasPermission(permission);
}
```

Tho this requires some refactoring (and an api tick?).

**Comment by bartico6** [comment]
You overcomplicated it.
```var args = new PlayerPermissionEventArgs(player, permission);
PlayerPermission(args);
return args.Handled```
is enough. It'll either give back a null when a hook returns null, or the outcome.
Then you just check the hook outcome if it's null (and execute logic) or deny/allow based on the boolean non-null value.

**Comment by deadsurgeon42** [comment]
But `HandledEventArgs.Handled` is `bool` not `bool?` ?

**Comment by bartico6** [comment]
Why not make it one?

**Comment by bartico6** [comment]
Or even better, make Handled an enum in those "handled in different ways" cases.

```c#
public enum PermissionHandleType {
UNHANDLED = 0,
DENIED = 1,
GRANTED = 2
}
```
Unless someone has an (unreasonable, in my opinion) distaste for enums, this could be very self-explanatory. Handled as a member of the args object should however be renamed and documented so it's more obvious it's not a "block it" trigger like it's for almost all other hooks.

**Comment by deadsurgeon42** [comment]
Enum is fine but we can't make Handled a `bool?` because it's from `HandledEventArgs` which is a .net framework type.

**Comment by ProfessorXZ** [comment]
Someone should prepare a PR for this.

**Comment by ProfessorXZ** [comment]
Someone like @Enerdy <3

**Comment by deadsurgeon42** [comment]
Ohhhh I can do a pr, just wanted to confirm before working on it.

**Comment by deadsurgeon42** [comment]
https://github.com/Pryaxis/TShock/pull/1476

**Comment by ProfessorXZ** [comment]
Implemented in #1476
